torch/tensor.py

```diff
@@ -313,23 +354,39 @@ def share_memory_(self):
         This is a no-op if the underlying storage is already in shared memory
         and for CUDA tensors. Tensors in shared memory cannot be resized.
         """
+        relevant_args = (self,)
+        from torch._overrides import has_torch_function, handle_torch_function
+        if type(self) is not Tensor and has_torch_function(relevant_args):
+            return handle_torch_function(Tensor.share_memory_, relevant_args, self)
         self.storage().share_memory_()
         return self
 
     def __reversed__(self):
         r"""Reverses the tensor along dimension 0."""
+        relevant_args = (self,)
+        from torch._overrides import has_torch_function, handle_torch_function
+        if type(self) is not Tensor and has_torch_function(relevant_args):
+            return handle_torch_function(Tensor.__reversed__, relevant_args, self)
         if self.dim() == 0:
             return self
         else:
             return self.flip(0)
 
     def norm(self, p="fro", dim=None, keepdim=False, dtype=None):
```

Comment:
The modifications in this file raise an interesting design problem which we didn't really discuss in the RFC, which we should probably talk about now. A lot of functions are duplicated as methods, as you can see here (where Tensor.norm is just another way of saying torch.norm). The choice you've made with these modifications is that the method is overrideable independently of the function (so a subclass could decide to define torch.norm and Tensor.norm differently). This is the maximally flexible choice. But if I override `torch.norm` but not `Tensor.norm`, what happens in this situation? What about vice versa? If I understand how the code here is written, the former will work, but the latter will not. You are leaking an internal implementation detail (methods are implemented by calling functions) into the public API. Should we perhaps define some rule where you ONLY define one of these, and it applies in both cases? (So for example, Tensor.norm would simply be unoverrideable, you only override torch.norm)? How could we communicate this to users?

I feel a lot of the problems here stem from the fact that we have manually implemented some Python methods on Tensor, while others are direct bindings from C++. In the ideal world, everything would be bound from C++ and so there would have been zero diff in this file. Does it make things simpler if we port all of these methods to C++ before attempting this PR? Can we make sure it remains *possible* to do this port at some later point in time? (So e.g., compare the behavior of Tensor.norm versus Tensor.__abs__)

Reply:
> The modifications in this file raise an interesting design problem which we didn't really discuss in the RFC, which we should probably talk about now. A lot of functions are duplicated as methods, as you can see here (where Tensor.norm is just another way of saying torch.norm). The choice you've made with these modifications is that the method is overrideable independently of the function (so a subclass could decide to define torch.norm and Tensor.norm differently). This is the maximally flexible choice. But if I override `torch.norm` but not `Tensor.norm`, what happens in this situation? What about vice versa? If I understand how the code here is written, the former will work, but the latter will not. You are leaking an internal implementation detail (methods are implemented by calling functions) into the public API. Should we perhaps define some rule where you ONLY define one of these, and it applies in both cases? (So for example, Tensor.norm would simply be unoverrideable, you only override torch.norm)? How could we communicate this to users?

This is all more-or-less true, except that methods don't always exist in the `torch` namespace, and vice versa. Also, I found one or two that have a different behavior, so instead of causing confusion and so on, I decided to go for the "maximally flexible" strategy.

> I feel a lot of the problems here stem from the fact that we have manually implemented some Python methods on Tensor, while others are direct bindings from C++. In the ideal world, everything would be bound from C++ and so there would have been zero diff in this file.

This is true.

> Does it make things simpler if we port all of these methods to C++ before attempting this PR?

It probably doesn't unless we move them to ATen and don't manually bind them.

> Can we make sure it remains possible to do this port at some later point in time? (So e.g., compare the behavior of `Tensor.norm` versus `Tensor.__abs__`) [sic]

That is a different issue entirely. Unfortunately the name/function we get passed into the public API comes from the argument parser, and since `Tensor.__abs__` and `Tensor.abs` both point to the same low-level C++ method, it becomes impossible to determine which route was taken from inside. So `Tensor.abs` is passed instead in both cases to `__torch_function__`. This is related to the issue of `Tensor.__add__` vs `Tensor.add`. I believe we discussed this in our call as well.

In Python, the issue is similar. `func` is the same for a number of dunder/normal methods, and so it becomes impossible to distinguish.

When moving to C++, if we separate `__add__` and `add` and so on, then the issue goes away.

Reply:
I've documented this behavior.